### PR TITLE
Support wagtail42

### PR DESCRIPTION
**Pull request by nickmoreton**
This updates the package for Wagtail 4.2

Drops: support for Wagtail < 4.1

For completeness I have merged in the following previous work:

- https://github.com/wagtail-nest/wagtail-ab-testing/pull/48
- https://github.com/wagtail-nest/wagtail-ab-testing/pull/49

### CODE REVIEW

**Review by victoriachan — CHANGES_REQUESTED** [review]
Hi Nick, there are a couple of mentions of wagtail 3 in Readme and setup to remove. They might have got merged in when you merged the other MRs. Also when testing locally, I got this error: `ValueError: Related model 'wagtailcore.pagerevision' cannot be resolved`. Looks like one of the migration files needs updating.

**Comment by nickmoreton** [comment]
> Also when testing locally, I got this error: `ValueError: Related model 'wagtailcore.pagerevision' cannot be resolved`. Looks like one of the migration files needs updating.

Thanks this is now resolved in: https://github.com/wagtail-nest/wagtail-ab-testing/pull/51/commits/3c63ca4eb1b6709ff16e9f8fdd9fe738aae59841

File: setup.py [hunk 1]
```diff
@@ -34,15 +34,13 @@
...
         "Framework :: Django",
-        "Framework :: Django :: 3.0",
         "Framework :: Django :: 3.2",
         "Framework :: Django :: 4.0",
         "Framework :: Django :: 4.1",
         "Framework :: Wagtail",
-        "Framework :: Wagtail :: 2",
-        "Framework :: Wagtail :: 3.0",
-        "Framework :: Wagtail :: 4.0",
+        "Framework :: Wagtail :: 4",
```
Comment (nickmoreton) [inline setup.py:43]: This wasn't right either. I tend to follow the lead from [wagtailmedia package](https://github.com/torchbox/wagtailmedia/blob/cafdf367b0de76c2d6a1c6bf91d2fc57a153f77b/setup.py#L54)

File: wagtail_ab_testing/test/__init__.py [hunk 2]
```diff
@@ -1 +1,4 @@
-default_app_config = "wagtail_ab_testing.test.apps.WagtailAbTestingTestAppConfig"
+from django import VERSION as DJANGO_VERSION
```
[the remark this answers is not in the document]
Comment (nickmoreton) [reply wagtail_ab_testing/test/__init__.py]: Resolved in https://github.com/wagtail-nest/wagtail-ab-testing/pull/51/commits/391be2e429cce11cadf4b730f93c2db00d2c2548

File: wagtail_ab_testing/__init__.py [hunk 3]
```diff
@@ -1 +1,4 @@
-default_app_config = "wagtail_ab_testing.apps.WagtailAbTestingAppConfig"
+from django import VERSION as DJANGO_VERSION
+
+if DJANGO_VERSION < (3, 2):
```
[the remark this answers is not in the document]
Comment (nickmoreton) [reply wagtail_ab_testing/__init__.py]: https://github.com/wagtail-nest/wagtail-ab-testing/pull/51/commits/391be2e429cce11cadf4b730f93c2db00d2c2548

File: tox.ini [hunk 4]
```diff
@@ -17,21 +19,23 @@ commands = coverage run testmanage.py test --deprecation all
...
     dj32: Django>=3.2,<3.3
+    dj40: Django>=4.0,<4.1
+    dj41: Django>=4.1,<4.2
     djmain: git+https://github.com/django/django.git@main#egg=Django
 
-    wa211: wagtail>=2.11,<2.12
-    wa215: wagtail>=2.15rc1,<2.16
-    wa216: wagtail>=2.16.1,<2.17
+    wa41: wagtail>=4.1,<4.2
+    wa42: wagtail>=4.2,<5.0
```
[the remark this answers is not in the document]
Comment (nickmoreton) [reply tox.ini:35]: https://github.com/wagtail-nest/wagtail-ab-testing/pull/51/commits/391be2e429cce11cadf4b730f93c2db00d2c2548

File: setup.py [hunk 5]
```diff
@@ -33,22 +33,25 @@
...
+        "Programming Language :: Python :: 3.10",
         "Framework :: Django",
-        "Framework :: Django :: 2.2",
         "Framework :: Django :: 3.0",
         "Framework :: Django :: 3.2",
+        "Framework :: Django :: 4.0",
+        "Framework :: Django :: 4.1",
         "Framework :: Wagtail",
         "Framework :: Wagtail :: 2",
+        "Framework :: Wagtail :: 3.0",
```
[the remark this answers is not in the document]
Comment (nickmoreton) [reply setup.py]: Resolved in https://github.com/wagtail-nest/wagtail-ab-testing/pull/51/commits/391be2e429cce11cadf4b730f93c2db00d2c2548

File: setup.py [hunk 6]
```diff
@@ -33,22 +33,25 @@
         "Programming Language :: Python :: 3.7",
         "Programming Language :: Python :: 3.8",
         "Programming Language :: Python :: 3.9",
+        "Programming Language :: Python :: 3.10",
         "Framework :: Django",
-        "Framework :: Django :: 2.2",
         "Framework :: Django :: 3.0",
```
[the remark this answers is not in the document]
Comment (nickmoreton) [reply setup.py]: Resolved in https://github.com/wagtail-nest/wagtail-ab-testing/pull/51/commits/391be2e429cce11cadf4b730f93c2db00d2c2548

File: setup.py [hunk 7]
```diff
@@ -33,22 +33,25 @@
         "Programming Language :: Python :: 3.7",
         "Programming Language :: Python :: 3.8",
         "Programming Language :: Python :: 3.9",
+        "Programming Language :: Python :: 3.10",
```
[the remark this answers is not in the document]
Comment (nickmoreton) [reply setup.py:36]: Resolved in https://github.com/wagtail-nest/wagtail-ab-testing/pull/51/commits/391be2e429cce11cadf4b730f93c2db00d2c2548

File: README.md [hunk 8]
```diff
@@ -18,6 +18,10 @@ Key features:
 
 ## Usage
 
+**The code examples below assume you are using Wagtail v3.0+** If you need information for an earlier version of Wagtail see the [Release notes](https://docs.wagtail.org/en/stable/releases/upgrading.html) for guidance.
```
[the remark this answers is not in the document]
Comment (nickmoreton) [reply README.md]: Resolved in https://github.com/wagtail-nest/wagtail-ab-testing/pull/51/commits/391be2e429cce11cadf4b730f93c2db00d2c2548

**Comment by nickmoreton** [comment]
Thanks @victoriachan sorry for missing so much here. They should all be resolved now.

File: wagtail_ab_testing/models.py [hunk 9]
```diff
@@ -77,7 +73,7 @@ class AbTest(models.Model):
     page = models.ForeignKey('wagtailcore.Page', on_delete=models.CASCADE, related_name='ab_tests')
     name = models.CharField(max_length=255)
     hypothesis = models.TextField(blank=True)
-    variant_revision = models.ForeignKey(get_revision_model(), on_delete=models.CASCADE, related_name='+')
+    variant_revision = models.ForeignKey('wagtailcore.Revision', on_delete=models.CASCADE, related_name='+')
```
Comment (nickmoreton) [inline wagtail_ab_testing/models.py:76]: Actually there was more that could be done here + a new migration.

This should fix sites that have the package already installed but the `treatment_revision` foreign key needs to be re-mapped as well.

Comment (nickmoreton) [reply wagtail_ab_testing/models.py:76]: I'm having second thoughts about this change. It probably needs more work to update the variant_revision fields. Like `rebuild_references_index` command in wagtail.

**Comment by pymike00** [comment]
Thank you for this, @nickmoreton.

Much appreciated, gonna give it a try! 

Do you actually work at Torchbox?

I wonder, with all due respect to the repo's owner and their time, why wasn't this PR even considered for merging? 

Did they abandon the project?

Best,
M.

**Comment by mwesterhof** [comment]
i'm with @pymike00 on this one, would love to know if there's been any changes regarding the maintenance of this project. We're currently investigating options for adding A/B testing support to some wagtail projects, but the lack of support for newer WT versions is quite a deal breaker for us. 

Is this something either of you knows the answer to, @kaedroho or @victoriachan ?

File: wagtail_ab_testing/models.py (hunk 9 above)
Comment (Stormheg) [reply wagtail_ab_testing/models.py:76]: I agree with you @nickmoreton that this is not enough to migrate to the `Revision` model. I'm investigating the best way to go about this.

**Review by Stormheg — COMMENTED** [review]
Wow, I can see quite the effort went into this. Thanks so much @nickmoreton ❤️ 

From a code-only perspective, almost all changes make sense to me... but I'm having difficulty testing this PR because I don't have an existing project (+ database) that uses this package to test this with. Because of this, I'm not going to approve this PR right now. Further (extensive) testing is required to make sure all works like expected.

Especially the migration from `PageRevision` to `Revision` is something that warrants further investigation to ensure it works for existing databases and new databases. Having an existing database to test with would be invaluable in testing.

I further suggest that we drop Python 3.7 support. Python 3.7 will be end-of-life next month.

**Comment by nitinb99** [comment]
We are thrilled about incorporating the wagtail-ab-testing package into our project. However, since it doesn't currently support version 4.2, we are unable to use it. We appreciate the effort you have put into adding support for the 4.2 version. We hope that this pull request gets merged swiftly, and we eagerly await its integration. Thank you to @nickmoreton, @Stormheg and @victoriachan   for your contributions.

File: wagtail_ab_testing/models.py (hunk 9 above)
Comment (Stormheg) [reply wagtail_ab_testing/models.py:76]: I've implemented a fix for this in #52 by squashing migrations. See the explanation in the PR description.

**Comment by Stormheg** [comment]
Hey all,

I've spent a couple more hours on this and have come up with a revised PR (#52) that includes a fix for the PageRevision to Revision migration. This fix should work for new projects and projects already wagtail ab testing.

I don't think there are any other obstacles that prevent supporting Wagtail 4.1 and 4.2, hope my PR is merged soon 🤞

**Comment by Stormheg** [comment]
superseded by #52

**Pull request closed without merging**